File: python/creational/01_singleton.py
"""
设计模式: 单例模式 (Singleton)
分类: 创建型

意图: 保证一个类只有一个实例，并提供全局访问点。
核心思想: 重写 __new__ 控制对象创建，线程安全加锁。
编程范式: OOP — metaclass / __new__ 元编程控制实例化。
"""

import threading


class SingletonMeta(type):
    """线程安全的单例元类"""
    _instances: dict = {}
    _lock: threading.Lock = threading.Lock()

    def __call__(cls, *args, **kwargs):
        with cls._lock:
            if cls not in cls._instances:
                instance = super().__call__(*args, **kwargs)
                cls._instances[cls] = instance
        return cls._instances[cls]


class Database(metaclass=SingletonMeta):
    def __init__(self) -> None:
        self.connection = "connected"
        print("[DB] connected")

    def query(self, sql: str) -> None:
        print(f"[DB] query: {sql}")


# ── 演示 ──────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    db1 = Database()
    db2 = Database()

    db1.query("SELECT * FROM users")
    db2.query("INSERT INTO logs VALUES (1)")

    print(f"same instance: {db1 is db2}")

File: cpp/creational/01_singleton.cpp
/**
 * 设计模式: 单例模式 (Singleton)
 * 分类: 创建型
 *
 * 意图: 保证一个类只有一个实例，并提供全局访问点。
 * 核心思想: 私有构造函数 + 静态实例 + 线程安全的懒汉式初始化。
 * 编程范式: OOP — 封装、静态成员控制实例生命周期。
 */

#include <iostream>
#include <mutex>
#include <string>

// ── 线程安全的 Meyer's Singleton ──────────────────────────────────────────
class Database {
public:
    // 全局访问点：局部静态变量天然线程安全（C++11 起）
    static Database& instance() {
        static Database db;
        return db;
    }

    void query(const std::string& sql) const {
        std::cout << "[DB] query: " << sql << "\n";
    }

    // 禁止拷贝与赋值
    Database(const Database&)            = delete;
    Database& operator=(const Database&) = delete;

private:
    Database() { std::cout << "[DB] connected\n"; }
    ~Database() { std::cout << "[DB] disconnected\n"; }
};

// ── 演示 ──────────────────────────────────────────────────────────────────
int main() {
    Database& db1 = Database::instance();
    Database& db2 = Database::instance();

    db1.query("SELECT * FROM users");
    db2.query("INSERT INTO logs VALUES (1)");

    std::cout << "same instance: " << std::boolalpha << (&db1 == &db2) << "\n";
    return 0;
}
